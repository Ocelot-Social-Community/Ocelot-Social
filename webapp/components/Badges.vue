<template>
  <div :class="[badges.length === 2 && 'hc-badges-dual']" class="hc-badges">
    <div v-for="badge in badges" :key="badge.id" class="hc-badge-container">
      <img :title="badge.key" :src="badge.icon | proxyApiUrl" class="hc-badge" />
    </div>
  </div>
</template>

<script>
export default {
  props: {
    badges: {
      type: Array,
      default: () => [],
    },
  },
}
</script>

<style lang="scss">
.hc-badges {
  text-align: center;
  position: relative;

  .hc-badge-container {
    display: inline-block;
    position: unset;
    overflow: hidden;
    vertical-align: middle;
  }

  .hc-badge {
    display: block;
    width: 100%;
  }

  $size: 30px;
  $offset: $size * -0.2;

  &.hc-badges-dual {
    padding-top: $size / 2 - 2;
  }

  .hc-badge-container {
    width: $size;
    height: 26px;
    margin-left: -1px;

    &:nth-child(3n - 1) {
      margin-top: -$size - $offset - 3;
      margin-left: -$size * 0.33 - $offset - 2;
    }
    &:nth-child(3n + 0) {
      margin-top: $size + $offset + 3;
      margin-left: -$size;
    }
    &:nth-child(3n + 1) {
      margin-left: -6px;
    }
    &:first-child {
      margin-left: -$size / 3;
    }
    &:last-child {
      margin-right: -$size / 3;
    }
  }
}
</style>
