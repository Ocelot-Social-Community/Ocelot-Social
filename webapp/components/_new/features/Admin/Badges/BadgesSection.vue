<template>
  <div class="badge-section">
    <h4>{{ title }}</h4>
    <div class="badge-container">
      <button
        v-for="badge in badges"
        :key="badge.id"
        @click="toggleBadge(badge)"
        :class="{ badge, inactive: !badge.isActive }"
      >
        <img :src="badge.icon" :alt="badge.description" />
      </button>
    </div>
  </div>
</template>

<script>
export default {
  props: {
    title: {
      type: String,
    },
    badges: {
      type: Array,
    },
  },
  methods: {
    toggleBadge(badge) {
      this.$emit('toggleBadge', badge)
    },
  },
}
</script>

<style scoped>
.badge-section h4 {
  margin-bottom: 24px;
}
.badge-container {
  display: flex;
  margin-bottom: 36px;
  flex-flow: row wrap;
  gap: 10px;
}
.badge:hover {
  cursor: pointer;
}
.badge img {
  width: 70px;
}
.badge.inactive {
  opacity: 0.3;
  filter: grayscale(0.4);
}
</style>
