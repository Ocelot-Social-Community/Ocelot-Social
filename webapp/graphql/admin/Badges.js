import gql from 'graphql-tag'

export const queryBadges = () => gql`
  query {
    Badge {
      id
      type
      icon
    }
  }
`

export const verify = () => gql`
  mutation ($badgeId: ID!, $userId: ID!) {
    verify(badgeId: $badgeId, userId: $userId) {
      id
      verified {
        id
      }
      badges {
        id
      }
    }
  }
`

export const reward = () => gql`
  mutation ($badgeId: ID!, $userId: ID!) {
    reward(badgeId: $badgeId, userId: $userId) {
      id
      verified {
        id
      }
      badges {
        id
      }
    }
  }
`

export const unreward = () => gql`
  mutation ($badgeId: ID!, $userId: ID!) {
    unreward(badgeId: $badgeId, userId: $userId) {
      id
      verified {
        id
      }
      badges {
        id
      }
    }
  }
`
