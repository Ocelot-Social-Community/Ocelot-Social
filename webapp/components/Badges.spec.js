import { shallowMount } from '@vue/test-utils'
import Badges from './Badges.vue'

describe('Badges.vue', () => {
  let propsData

  beforeEach(() => {
    propsData = {}
  })

  describe('shallowMount', () => {
    const Wrapper = () => {
      return shallowMount(Badges, { propsData })
    }

    it('has class "hc-badges"', () => {
      expect(Wrapper().find('.hc-badges').exists()).toBe(true)
    })

    describe('given a badge', () => {
      beforeEach(() => {
        propsData.badges = [{ id: '1', icon: '/path/to/some/icon' }]
      })

      it('proxies badge icon, which is just a URL without metadata', () => {
        expect(Wrapper().find('img[src="/api/path/to/some/icon"]').exists()).toBe(true)
      })
    })
  })
})
