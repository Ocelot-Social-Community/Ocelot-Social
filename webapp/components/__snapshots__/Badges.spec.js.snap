// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Badges.vue with badges in presentation mode renders 1`] = `
<div>
  <div
    class="hc-badges"
    scale="1.2"
  >
    <div
      class="hc-badge-container"
    >
      <img
        class="hc-badge"
        src="/api/path/to/some/icon"
        title="Some description"
      />
    </div>
    <div
      class="hc-badge-container"
    >
      <img
        class="hc-badge"
        src="/api/path/to/another/icon"
        title="Another description"
      />
    </div>
    <div
      class="hc-badge-container"
    >
      <img
        class="hc-badge"
        src="/api/path/to/third/icon"
        title="Third description"
      />
    </div>
  </div>
</div>
`;

exports[`Badges.vue with badges in selection mode clicking on second badge selects badge 1`] = `
<div>
  <div
    class="hc-badges"
    scale="1.2"
  >
    <div
      class="hc-badge-container"
    >
      <img
        class="hc-badge"
        src="/api/path/to/some/icon"
        title="Some description"
      />
    </div>
    <button
      class="hc-badge-container selectable selected"
    >
      <img
        class="hc-badge"
        src="/api/path/to/another/icon"
        title="Another description"
      />
    </button>
    <button
      class="hc-badge-container selectable"
    >
      <img
        class="hc-badge"
        src="/api/path/to/third/icon"
        title="Third description"
      />
    </button>
  </div>
</div>
`;

exports[`Badges.vue with badges in selection mode clicking twice on second badge deselects badge 1`] = `
<div>
  <div
    class="hc-badges"
    scale="1.2"
  >
    <div
      class="hc-badge-container"
    >
      <img
        class="hc-badge"
        src="/api/path/to/some/icon"
        title="Some description"
      />
    </div>
    <button
      class="hc-badge-container selectable"
    >
      <img
        class="hc-badge"
        src="/api/path/to/another/icon"
        title="Another description"
      />
    </button>
    <button
      class="hc-badge-container selectable"
    >
      <img
        class="hc-badge"
        src="/api/path/to/third/icon"
        title="Third description"
      />
    </button>
  </div>
</div>
`;

exports[`Badges.vue with badges in selection mode renders 1`] = `
<div>
  <div
    class="hc-badges"
    scale="1.2"
  >
    <div
      class="hc-badge-container"
    >
      <img
        class="hc-badge"
        src="/api/path/to/some/icon"
        title="Some description"
      />
    </div>
    <button
      class="hc-badge-container selectable"
    >
      <img
        class="hc-badge"
        src="/api/path/to/another/icon"
        title="Another description"
      />
    </button>
    <button
      class="hc-badge-container selectable"
    >
      <img
        class="hc-badge"
        src="/api/path/to/third/icon"
        title="Third description"
      />
    </button>
  </div>
</div>
`;

exports[`Badges.vue without badges renders in presentation mode 1`] = `
<div>
  <div
    class="hc-badges"
  />
</div>
`;

exports[`Badges.vue without badges renders in selection mode 1`] = `
<div>
  <div
    class="hc-badges"
  />
</div>
`;
